### PR TITLE
Τimezone support

### PR DESCRIPTION
**Pull request by kloukas**
Added support for timezones, both on instance and runtime level. Also adjusted some tests to use providers and handle timezones.

### CODE REVIEW

**Comment by kloukas** [comment]
I'm not sure how to send PHPUnit and coverage information, I'll paste the output below. Let me know if there's a better way to do it.

```
PHPUnit 8.5.4 by Sebastian Bergmann and contributors.

................................................................. 65 / 70 ( 92%)
.....                                                             70 / 70 (100%)

Time: 1.99 seconds, Memory: 8.00 MB

OK (70 tests, 280 assertions)

Generating code coverage report in Clover XML format ... done [146 ms]

Generating code coverage report in HTML format ... done [181 ms]

Generating code coverage report in PHP format ... done [7 ms]


Code Coverage Report:    
  2020-05-16 11:03:44    
                         
 Summary:                
  Classes:  0.00% (0/5)  
  Methods: 47.62% (10/21)
  Lines:   71.70% (38/53)

\CodeIgniter\Tasks::CodeIgniter\Tasks\CronExpression
  Methods:  83.33% (10/12)   Lines:  95.00% ( 38/ 40)
```

File: src/CronExpression.php [hunk 1]
```diff
@@ -18,13 +18,52 @@ class CronExpression
...
+     * Allows us to set global timezone for all tasks
+     * on construct
+     * 
+     * @param string $timezone
+     * 
+     * @return void
+     */
+    public function __construct(string $timezone = null)
+    {
+        if($timezone === null){
```
Comment (MGatner) [inline src/CronExpression.php]: The goal is for this to make it into `codeigniter4/codeigniter4` eventually so should follow the same style guide:
```
if ($timezone === null)
{
```

Comment (MGatner) [reply src/CronExpression.php]: Same thing throughout these changes.

File: src/CronExpression.php [hunk 2]
```diff
@@ -18,13 +18,52 @@ class CronExpression
...
+    /**
+     * Checks whether cron expression should be run. Allows
+     * for custom timezone to be used for specific task
+     * 
+     * @param string $expression
+     * @param string $timezone
+     * 
+     * @return void
+     */
+    public function shouldRun(string $expression, $timezone = null): bool
```
Comment (MGatner) [inline src/CronExpression.php]: `string $timezone = null`

**Review by MGatner — CHANGES_REQUESTED** [review]
I like this change especially for distributable projects that may need to manage tasks for multiple timezones and locales. I mentioned a few style issues but there are a lot that would need to comply with the framework guide. Please review https://github.com/codeigniter4/CodeIgniter4/blob/develop/contributing/styleguide.rst

**Comment by kloukas** [comment]
I ran phpcbf on the files using the ruleset in codeigniter4/coding-standard. Let me know if that fixed the issues. Do you think it might be helpfull if we added that to the composer file and created the same pre-commit hook the main repository has to autofix issues?

File: src/CronExpression.php [hunk 3]
```diff
@@ -4,155 +4,204 @@
...
+
+	/**
+	 * Sets the timezone that dates should be examined under.
+	 *
+	 * @param string $timezone
+	 *
+	 * @return $this
+	 * @throws \Exception Thrown if $timezone is not a valid timezone string.
+	 */
+	public function timezone(string $timezone)
```
Comment (lonnieezell) [inline src/CronExpression.php]: please rename to `setTimezone` for clarity. Also - I think the logic you added into `shouldRun()` dealing with timezones could be placed here, allowing `shouldRun` to focus on it's one task, and to remove extra arguments.

**Comment by lonnieezell** [comment]
Nice addition, thanks! I just had the one comment that was just a preference, nothing wrong with the way you did it. 

Adding coding_standard and the same script we use for the CI core repo makes a lot of sense.

File: src/CronExpression.php (hunk 3 above)
Comment (kloukas) [reply src/CronExpression.php]: Do you mean lines 56-66? If we move that to setTimezone we would lose the ability to add custom timezones at run time. Doing something like
```
$cron = new CronExpression();
$cron ->setTimezone('America/Chicago');
$cron->shouldRun('* 10 * * *'); // Checks Chicago Time
$cron->shouldRun('* 10 * * *','Europe/Athens'); // Checks Athens
$cron->shouldRun('* 10 * * *'); // Checks Chicago again
```

would need setTimezone() to be called anytime we needed to switch timezones. If you don't think such a functionality is not needed I can move all that logic to setTimezone, sure.

Comment (lonnieezell) [reply src/CronExpression.php]: It was late last night when I looked at this, and my coffee hasn't fully kicked in yet this morning :) but let me see if I can understand. The default timezone should be `Config\App::$appTimezone`, right? I guess we're not taking that into account, currently. When I was looking at it lastnight, I was thinking it would be used like? 

```
$cron = new CronExpression();
$cron->setTimezone('Europe/Athens')->shouldRun(...);
```

Basically, I was trying to minimize argument usage since I feel that's clearer, but as I think through it I think what you've done here is fine. It allows us to pass in the base timezone as a dependency through `setTimezone` which makes testing easier anyway, and the additional argument makes sense. Besides, this class is an internal file, anyway. 

Long story short: ignore me lol

Comment (kloukas) [reply src/CronExpression.php]: No worries. Minimizing arguments is not an invalid argument.
The default timezone should be Config\App::$appTimezone, you're right. I used app_timezone() instead, I'll fix it.
(EDIT: app_timezone() apparently fetches Config\App::$appTimezone, is it better to use that or go directly for the config class?)

Regarding the shouldRun() function, I think the main thing we need to figure out is how the CronExpression class is going to be used and how it will be connected to the Tasks.

Is it going to be simply a "helper" class?
For example if Tasks is something like
```php
class Task{

    public function __construct($job,$expression,$timezone){
        //
        // Pretend we handle args here
        //

        $this->cron = new CronExpression();
        $this->cron->setTimezone($this->timezone);
    }

    public function run(){
        if($this->cron->shouldRun($this->expression)){
            // Do Job
        }
    }

    // Whatever other logic is needed
}


```

then setting a timezone at shouldRun should not be needed.

If instead we want to do something more complicated, like assign each task multiple timezones, then it might be helpful to be able to assign the timezone on shouldRun().

That being said, I think it's a really extreme edge case that a Task would need two different timezones. I can't honestly think of any examples where that would be needed. I'll move everything to setTimezone and fix the tests.

**Comment by kloukas** [comment]
I think that addresses most of the changes. Let me know if you think anything else needs to be fixed.
I'll make another PR for the coding standards change in a bit.

File: src/CronExpression.php (hunk 3 above)
Comment (lonnieezell) [reply src/CronExpression.php]: I had forgotten you were using app_timezone() already. No need to change. 

The intent is that this class is only used internally and not exposed publicly. I cannot think of a reason we would ever need multiple timezones for a single task. Not something I want to support either :) On the weird case they actually come up with a use for that, they can break it up into multiple jobs lol. 

I'll try to get some basic docs written up very soon so that we know how the pieces should fit together.
